definition_validator/validator.cc

```diff
@@ -758,7 +758,7 @@ void validateUnsatisfiableRequiredAncestors(core::Context ctx, const core::Class
             requiredClasses.emplace_back(ancst);
         }
 
-        if (ancst.symbol.data(ctx)->typeArity(ctx) > 0) {
+        if (ancst.symbol.data(ctx)->typeArity(ctx) > 0 && !ancst.symbol.data(ctx)->isSingletonClass(ctx)) {
```

Comment:
Could we add a test for this?

Reply:
In particular, this code is buggy. We cannot allow generics for `T.class_of` for the same reason that we do not for instance classes. The code as written suffers from this bug:

#### Input

[→ View on sorbet.run](https://sorbet.run/?arg=--enable-experimental-requires-ancestor#%23%20typed%3A%20strict%0Aextend%20T%3A%3ASig%0A%0Aclass%20Parent%0A%20%20class%20%3C%3C%20self%0A%20%20%20%20extend%20T%3A%3ASig%0A%20%20%20%20extend%20T%3A%3AGeneric%0A%0A%20%20%20%20Elem%20%3D%20type_member%0A%0A%20%20%20%20sig%20%7Breturns%28T.nilable%28Elem%29%29%7D%0A%20%20%20%20attr_accessor%20%3Aelem%0A%20%20end%0A%20%20%0Aend%0A%0Amodule%20Example%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%0A%20%20requires_ancestor%20%7BT.class_of%28Parent%29%7D%0A%0A%20%20sig%20%7Bvoid%7D%0A%20%20def%20example%0A%20%20%20%20result%20%3D%20self.elem%0A%20%20%20%20T.reveal_type%28result%29%0A%20%20end%0Aend)

```ruby
# typed: strict
extend T::Sig

class Parent
  class << self
    extend T::Sig
    extend T::Generic

    Elem = type_member

    sig {returns(T.nilable(Elem))}
    attr_accessor :elem
  end
  
end

module Example
  extend T::Sig
  extend T::Helpers

  requires_ancestor {T.class_of(Parent)}

  sig {void}
  def example
    result = self.elem
    T.reveal_type(result)
  end
end
```

#### Observed output

```
editor.rb:25: Expression does not have a fully-defined type (Did you reference another class's type members?) https://srb.help/7011
    25 |    result = self.elem
                     ^^^^^^^^^
  Note:
    Historically, users seeing this error has represented finding a bug in Sorbet
    (or at least finding a test case for which there could be a better error message).
    Consider searching for similar bugs at https://github.com/sorbet/sorbet/issues
    or reporting a new one.

editor.rb:26: Revealed type: T.untyped https://srb.help/7014
    26 |    T.reveal_type(result)
            ^^^^^^^^^^^^^^^^^^^^^
  Got T.untyped originating from:
    editor.rb:25:
    25 |    result = self.elem
                     ^^^^^^^^^
Errors: 2
```

Reply:
We should special case something to deal with the fact that all singleton classes already have one type member with the name `core::Names::Constants::AttachedClass()`, but not permit a singleton class that has other type members defined on it.